### PR TITLE
[FSDP2] reset FSDPParam.sharded_param in lazy_init

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132954
* #133059

motivated by FSDP2 + DoRA https://github.com/pytorch/pytorch/issues/132721

after meta init, we need a user-defined function to move DoRALinear.magnitude from device=meta to device=cuda
The problem is how to trigger reset_sharded_param or _apply to update FSDPParam. Otherwise lazy_init complains that DoRALinear.magnitude are still on device=meta

credit to @awgu for chasing after DDP lazy_init to unblock the PR

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 1]
```diff
@@ -191,6 +191,12 @@ def _init_mp_dtypes(self) -> None:
 
     def lazy_init(self):
         # Lazy init should be idempotent
+        # User may initialize params again after meta init.
+        # For example, we initialize DoRA magnitudes base on
+        # other parameters loaded from the state dict.
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:196]: (IIUC, it is not specific to meta init or not.)

**Review by awgu — APPROVED** [review]
Thanks!

**Comment by weifengpy** [comment]
the CI error is relevant and I want to discuss my minimized repro with @awgu https://github.com/pytorch/pytorch/issues/132721

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 2]
```diff
@@ -191,6 +191,12 @@ def _init_mp_dtypes(self) -> None:
...
+        with torch.no_grad():
+            for fsdp_param in self.fsdp_params:
+                fsdp_param.reset_sharded_param()
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:199]: I think to fix https://github.com/pytorch/pytorch/issues/132978
```suggestion
        if self.is_sharded:
            with torch.no_grad():
                for fsdp_param in self.fsdp_params:
                    fsdp_param.reset_sharded_param()
```

Comment (weifengpy) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:199]: just tested and the gap is still in DDP(FSDP).  FSDPModule params are "unsharded" to `torch.Tensor` already before calling `.unshard`, but `FSDPParamGroup.is_sharded` is still `True` 

https://github.com/pytorch/pytorch/blob/a9954d22f89677fc22fb9764f1edb0818538478f/test/distributed/_composable/fsdp/test_fully_shard_comm.py#L1066

File: test/distributed/_composable/fsdp/test_fully_shard_init.py [hunk 3]
```diff
@@ -536,6 +536,41 @@ def test_fully_shard_multi_module_root(self):
...
+                self.layer1 = nn.Linear(3, 3, bias=False)
+                self.layer2 = nn.Linear(3, 3, bias=False)
+                self.weight_norm = nn.Parameter(torch.empty(3))
+
+            def init_weight_norm(self):
+                with torch.no_grad():
+                    weight_norm = torch.linalg.norm(
+                        self.layer1.weight, dim=1
+                    ) + torch.linalg.norm(self.layer2.weight, dim=1)
+                model.weight_norm = nn.Parameter(weight_norm)
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_init.py:553]: @ebsmothers  examples of DoRA.magnitude in case you want to remove hacky code

**Pull request closed without merging**